### PR TITLE
refactor: optimize calculate nodes.

### PR DESCRIPTION
**Pull request by domoritz**
In this PR, I am adding a new optimizer that (similar to the bin optimizer) merges calculate nodes at forks and moves them up in the dataflow (towards the source). I extended the calculate node to support multiple calculates in a single node.

### CODE REVIEW

File: examples/compiled/circle_github_punchcard.vg.json [hunk 1]
```diff
@@ -10,6 +10,11 @@
       "url": "data/github.csv",
       "format": {"type": "csv", "parse": {"time": "date"}},
       "transform": [
+        {
```
Comment (kanitw) [inline examples/compiled/circle_github_punchcard.vg.json]: This formula shouldn't get reordered?

File: src/compile/data/calculate.ts [hunk 2]
```diff
@@ -53,24 +51,64 @@ export class CalculateNode extends DataFlowNode {
...
+  /**
+   * Merge calculate nodes. This method either integrates the bin config from the other node
+   * or if this node already has a bin config, renames the corresponding signal in the model.
+   */
+  public merge(other: CalculateNode) {
+    for (const key of keys(other.calculates)) {
+      if (key in this.calculates) {
+        // make sure we are not merging something incompatible
+        if (this.calculates[key].calculate !== other.calculates[key].calculate) {
+          /* istanbul ignore next: This should never happen. */
```
Comment (kanitw) [inline src/compile/data/calculate.ts]: What if you have two layers that calculate two different things "as" the same variable?  

If it shouldn't happen, why add it in the code?

File: src/compile/data/optimizers.ts [hunk 3]
```diff
@@ -425,6 +403,60 @@ export class MergeBins extends BottomUpOptimizer {
...
 
+/**
+ * Merge calculate nodes and move them up through forks. Stop at filters and parse as we want them to stay before the calculate node.
+ *
+ * Similar to MergeBins but for Calculate Nodes.
+ */
+export class MergeCalculate extends BottomUpOptimizer {
+  public run(node: DataFlowNode): OptimizerFlags {
+    const parent = node.parent;
+    const moveCalcssUp = !(
```
Comment (kanitw) [inline src/compile/data/optimizers.ts]: `moveCalcsUp`  (Too many `s`'s)

File: src/compile/data/calculate.ts [hunk 4]
```diff
@@ -53,24 +51,64 @@ export class CalculateNode extends DataFlowNode {
...
-      type: 'formula',
-      expr: this.transform.calculate,
-      as: this.transform.as
-    };
+  public assemble(): VgFormulaTransform[] {
+    const transforms: VgFormulaTransform[] = [];
+
+    for (const {key, value} of entries(this.calculates)) {
+      if (key !== value.as) {
+        /* istanbul ignore next: This should never happen. */
```
Comment (kanitw) [inline src/compile/data/calculate.ts]: Why checking this if it shouldn't ever happen?  

(If you worry about bug, can you check this in a unit test instead? )

File: examples/compiled/waterfall_chart.vg.json [hunk 5]
```diff
@@ -88,7 +88,7 @@
       "transform": [
         {
           "type": "filter",
-          "expr": "datum[\"previous_sum\"] !== null && !isNaN(datum[\"previous_sum\"])"
+          "expr": "datum[\"sum_inc\"] !== null && !isNaN(datum[\"sum_inc\"])"
```
Comment (kanitw) [inline examples/compiled/waterfall_chart.vg.json]: This example's dataflow got swapped like crazy!

File: examples/compiled/rect_mosaic_labelled_with_offset.vg.json [hunk 6]
```diff
@@ -85,11 +85,8 @@
       "source": "source_0",
       "transform": [
         {
-          "type": "aggregate",
-          "groupby": ["Origin"],
-          "ops": ["min"],
-          "fields": ["xc"],
-          "as": ["min_xc"]
+          "type": "filter",
```
Comment (kanitw) [inline examples/compiled/rect_mosaic_labelled_with_offset.vg.json]: filter/aggregate shouldn't get swapped here.

File: src/compile/data/calculate.ts (hunk 4 above)
Comment (domoritz) [reply src/compile/data/calculate.ts]: This is an assertion that I added. It won't happen right now but I am worried that at some point I might forget to satisfy the assumptions.

**Comment by domoritz** [comment]
I think the logic is flawed. I should merge calculates like I merge parses, not like I merge bins. For bins we know that the keys are sufficient to differentiate what the bin calcualtes.

**Pull request closed without merging**